netket/utils/group/semigroup.py

```diff
@@ -0,0 +1,169 @@
+# pylint: disable=function-redefined
+
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+import itertools
+from typing import Callable, List
+
+from plum import dispatch
+
+import numpy as np
+
+from netket.utils import HashableArray
+from netket.utils.types import Array, DType, Shape
+
+
+class ElementBase(ABC):
+    @abstractmethod
+    def __call__(self, arg):
+        pass
+
+    def __matmul__(self, other):
+        return product(self, other)
+
+
+class Element(ElementBase):
+    pass
+
+
+@dataclass(frozen=True)
+class Identity(ElementBase):
+    def __call__(self, arg):
+        return arg
+
+    def __repr__(self):
+        return "Id()"
+
+
+@dispatch
+def product(a: Identity, _: Identity):
+    return a
+
+
+@dispatch
+def product(_: Identity, b: Element):
+    return b
+
+
+@dispatch
+def product(a: Element, _: Identity):
+    return a
+
+
+@dataclass(frozen=True)
+class Composite(Element):
+    left: Element
+    right: Element
+
+    def __call__(self, arg):
+        return self.left(self.right(arg))
+
+    def __repr__(self):
+        return f"{self.left} @ {self.right}"
+
+
+@dispatch
+def product(a: Element, b: Element):
+    return Composite(a, b)
+
+
+@dispatch
+def product(ab: Composite, c: Element):
+    bc = product(ab.right, c)
+    if isinstance(bc, Composite):
+        return Composite(ab, c)
+    else:
+        return Composite(ab.left, bc)
+
+
+@dispatch
+def product(a: Element, bc: Composite):
+    ab = product(a, bc.left)
+    if isinstance(ab, Composite):
+        return Composite(a, bc)
+    else:
+        return Composite(ab, bc.right)
+
+
+@dispatch
+def product(ab: Composite, cd: Composite):
+    bc = product(ab.right, cd.left)
+    if isinstance(bc, Composite):
+        return Composite(ab, cd)
+    else:
+        return Composite(ab.left, Composite(bc, cd.right))
+
+
+@dataclass(frozen=True)
+class NamedElement(Element):
+    name: str
+    action: Callable
+    info: str = ""
+
+    def __call__(self, arg):
+        return self.action(arg)
+
+    def __repr__(self):
+        return f"{self.name}({self.info})"
+
+
+@dataclass(frozen=True)
+class SemiGroup:
+    elems: List[Element]
+
+    def __post_init__(self):
+        # manually assign self.__hash == ... for frozen dataclass,
+        # see https://docs.python.org/3/library/dataclasses.html#frozen-instances
+        myhash = hash(tuple(hash(x) for x in self.elems))
+        object.__setattr__(self, "_SemiGroup__hash", myhash)
+
+    def __matmul__(self, other):
+        """
+        Direct product of this group with `other`.
+        """
+        return SemiGroup(
+            elems=[a @ b for a, b in itertools.product(self.elems, other.elems)],
+        )
+
+    def __call__(self, initial):
+        """
+        Apply all group elements to all entries of `initial` along the last axis.
+        """
+        initial = np.asarray(initial)
+        return np.array([np.apply_along_axis(elem, -1, initial) for elem in self.elems])
```

Comment:
I think that `__matmul__` should be implemented using dispatch, 'a la `ElementBase`. This allows us to do some tricks. For example, right now we cannot apply an element to a group even though this might make sense (eg, translate by a set amount the group of all rotations). I would use `product` here to represent the `matmul` of semigroups (and later we can add more things). It also allows to customise the product of groups/semigroups.

I think also here we are separating 2 operations that could be represented by the same operation, that is, `matmul`.
After all, when we do `group @ group` we want the current matmul, but if we do `group@ket` or `group@batchofket` we want `call`. there's no need to separate the two.